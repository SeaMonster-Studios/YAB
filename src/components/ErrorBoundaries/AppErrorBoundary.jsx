/* globals process */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
// import Raven from 'raven-js'

const styles = {
  container: {
    position: 'absolute',
    top: '50%',
    left: 0,
    right: 0,
    width: '100%',
    padding: 20,
    lineHeight: 1.5,
    fontSize: 16,
    backgroundColor: '#680000',
    color: '#fff',
    textAlign: 'center',
    fontWeight: 300,
    zIndex: 9999999999999999999999999999999999999999,
  },
  innerContainer: {
    maxWidth: 750,
    margin: '0 auto',
  },
  button: {
    textDecoration: 'underline',
    cursor: 'pointer',
    fontWeight: 300,
    color: '#fff',
    padding: 0,
    background: 'rgba(0,0,0,0)',
    border: 'none',
    fontSize: 16,
    lineHeight: 1.5,
  },
}

export default class AppErrorBoundary extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  }

  constructor(props) {
    super(props)

    this.state = {
      hasError: false,
    }
  }

  componentDidCatch(error, info) {
    this.setState(() => {
      return {
        hasError: true,
      }
    })

    if (
      process.env.NODE_ENV !== 'development' &&
      typeof document !== 'undefined'
    ) {
      // Raven.captureException(error, { extra: info })
    } else {
      console.error(
        'Error caught in Error AppErrorBoundary. This will reported to Sentry when not in development.',
      )
    }
  }

  render() {
    if (this.state.hasError) {
      return (
        <div style={styles.container} data-testid="app-error-boundary-error">
          <div style={styles.innerContainer}>
            <p>Sorry, something went wrong.</p>
            <ReportForm />
            <p>
              You can try reloading the page, or using the menu if it is
              available.
            </p>
          </div>
        </div>
      )
    }
    return this.props.children
  }
}

const ReportForm = () =>
  process.env.NODE_ENV !== 'development' && typeof document !== 'undefined' ? (
    <Form />
  ) : (
    <span />
  )

const Form = () => (
  <p data-testid="app-error-boundary-report-form-wrapper">
    This error has been reported to our development team, but{' '}
    <button
      // onClick={() => Raven.lastEventId() && Raven.showReportDialog()}
      style={styles.button}
    >
      click here to fill out a report.
    </button>
  </p>
)
