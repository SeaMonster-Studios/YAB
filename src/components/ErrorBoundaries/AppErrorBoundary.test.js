import React from 'react'
import { Simulate, render } from 'react-testing-library'
//
import AppErrorBoundary from './AppErrorBoundary'

const originalEnv = process.env.NODE_ENV

beforeAll(() => {
  process.env.NODE_ENV = 'production'
})

afterAll(() => {
  process.env.NODE_ENV = originalEnv
})

describe('AppErrorBoundary Component Test', () => {
  it('Renders children when there is no app error', () => {
    const { container, child } = renderSetup()

    // make sure it renders
    expect(container).toMatchSnapshot()

    expect(child).toBeDefined()
  })

  it('Renders error message and report form when there is an error', async () => {
    /* eslint-disable-next-line no-console */
    console.error = () => undefined

    const { container, getByTestId, child } = renderSetup()

    expect(child).toBeDefined()

    try {
      Simulate.click(child)
    } catch (error) {
      swallowError()
    }

    expect(getByTestId('app-error-boundary-error')).toBeDefined()
    expect(getByTestId('app-error-boundary-report-form-wrapper')).toBeDefined()
    expect(container).toMatchSnapshot()
  })
})

const swallowError = () => null

function renderSetup(overrides) {
  class Child extends React.Component {
    state = {
      errorMe: false,
    }
    render() {
      return (
        <span
          data-testid="app-error-boundary-test-child"
          onClick={() =>
            this.setState(prevState => ({
              ...prevState,
              errorMe: true,
            }))
          }
        >
          {/* eslint-disable-next-line react/jsx-no-undef */}
          {this.state.errorMe ? <IDontExist /> : <span />}
        </span>
      )
    }
  }

  const wrapper = render(
    <AppErrorBoundary>
      <Child />
    </AppErrorBoundary>,
  )
  const child = wrapper.getByTestId('app-error-boundary-test-child')

  return {
    ...wrapper,
    child,
    ...overrides,
  }
}
