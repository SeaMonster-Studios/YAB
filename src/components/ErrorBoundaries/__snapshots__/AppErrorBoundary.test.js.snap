// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppErrorBoundary Component Test Renders children when there is no app error 1`] = `
<div>
  <span
    data-testid="app-error-boundary-test-child"
  >
    <span />
  </span>
</div>
`;

exports[`AppErrorBoundary Component Test Renders error message and report form when there is an error 1`] = `
<div>
  <div
    data-testid="app-error-boundary-error"
    style="position: absolute; top: 50%; left: 0px; right: 0px; width: 100%; padding: 20px; line-height: 1.5; font-size: 16px; background-color: rgb(104, 0, 0); color: rgb(255, 255, 255); text-align: center; font-weight: 300; z-index: 1e+40;"
  >
    <div
      style="max-width: 750px; margin: 0px auto;"
    >
      <p>
        Sorry, something went wrong.
      </p>
      <p
        data-testid="app-error-boundary-report-form-wrapper"
      >
        This error has been reported to our development team, but
         
        <button
          style="text-decoration: underline; cursor: pointer; font-weight: 300; color: rgb(255, 255, 255); padding: 0px; background: rgba(0, 0, 0, 0); font-size: 16px; line-height: 1.5;"
        >
          click here to fill out a report.
        </button>
      </p>
      <p>
        You can try reloading the page, or using the menu if it is available.
      </p>
    </div>
  </div>
</div>
`;
